### ISSUE TITLE
Empty alt attribute added to images

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4551]
**Steps To Reproduce**

1. Create a post in a vanilla Backdrop installation.
2. Add an image to it, but leave the "Alternate text" optional field empty.
3. Save.
4. Inspect the image code in your browser ->

**Actual behavior**

`<img src="https://what.ever/files/styles/large/public/field/image/my-image.jpg" width="600" height="600" alt>`

or

`<img src="https://what.ever/files/styles/large/public/field/image/my-image.jpg" width="600" height="600" alt="">`

**Expected behavior**

The rendered HTML output should not include the `alt` attribute at all, if it was left empty.

**Comment by indigoxela** [comment]
> The rendered HTML output should not include the alt attribute at all, if it was left empty.

I disagree. The img tag *must* contain the alt attribute, empty or not. It is required.

https://www.w3schools.com/TAGS/att_img_alt.asp

**Comment by BWPanda** [comment]
Related? https://github.com/backdrop/backdrop-issues/issues/3006

**Comment by indigoxela** [comment]
> Related? #3006

Not really related. This issue here is about adding an attribute to a tag. The other one is about setting a form field to required by default.

**Comment by klonos** [comment]
I would like to start off by saying that there seems to be a lot of misconception around the usage of `alt`, and it is a fact that a lot of people have misinterpreted the standards as "use alt text at all times, at all costs", which is plain wrong. Posting articles on the web, expressing one's _opinions_ on the matter is OK; presenting those opinions as facts and/or standards is harmful...

> I disagree. The img tag *must* contain the alt attribute, empty or not. It is required.

As valuable a source of information as w3schools.com may be, it is NOT the authority in such matters. I'm happy to be proven wrong, but I have checked both https://www.w3.org/QA/Tips/altAttribute and https://www.w3.org/TR/WCAG10, and although it is advised/recommended to add an alt attribute, it is NOT _required_ by any standard I could find. The only relevant thing I managed to find is this:

> it is fine to have an empty alt attribute (e.g., `alt=""`)

`fine to have != must have;`

I will give you that, as mentioned in https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/F65, omitting the alt attribute or text alternative on img elements is a failure:

> If there is no source of text to provide an alternative for the image then assistive technologies are not able to identify the image or to convey its purpose to the user.

But this does not say anywhere that empty alt text should be added at all times. In fact, please read this which points to the opposite: https://www.w3.org/TR/WCAG20-TECHS/H37.html

> ...failing this test procedure does not necessarily mean that the success criterion has not been satisfied in some other way, only that this technique has not been successfully implemented and can not be used to claim conformance.

Furthermore (although not the same as `alt=""`), https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/F30 mentions that `alt=" "` is also a failure.

Also, https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/F39 mentions that:

> Providing a null text alternative (i.e., `alt=""`) will allow assistive technology to ignore the image and avoid a failure of this Success Criterion.
> ...
> ## Procedure
> 1. Identify any img elements that are used for decoration, spacing or other purpose that is not part of the meaningful content in the page
> 2. Check that the alt attribute for these elements is null.
>
> ## Expected Results
> If step #2 is false, this failure condition applies and the content fails the Success Criterion.

There are tools out there that will try to do image recognition, and use voice to announce the image contents (best AI guess, but still decent as the technology advances). By arbitrarily adding `alt=""`, we are instructing assistive technology to "skip" images as "decorative", when in fact they may be the main image of the article. This is plain wrong.

By adding an empty/null alt, we are just fooling ourselves ...perhaps also some WCAG checking tool/service to give us a nice result, without providing any real value though.

Interesting read on this matter (still NOT authoritative by any means - it clearly states that everything in there is "opinions"): https://juicystudio.com/wcag/altinhtml5.html. In that article, you will notice many "should"s, but nothing mentioning any hard requirement by a specific standard.

I personally agree with these 2 points, which apply to empty alt text:

> - Making the alt attribute required results in people providing nonsense values.
> - Making the alt attribute required does not mean it will be used correctly.

**Comment by indigoxela** [comment]
If we want to pass checks on https://validator.w3.org/, we have to use the alt attribute. A discussion about the real-world benefit might be sort of academic, because if Backdrop sites fail the validator, users will come nagging.

Validator screenshot if alt is missing:
![screenshot-w3c](https://user-images.githubusercontent.com/842066/90532542-aabb9780-e177-11ea-9443-ba033ad53aaa.png)

**Comment by klonos** [comment]
Thanks @indigoxela ...it does make sense from a "marketing" point.

I now see why they decided to make the alt text required in Drupal (not that it stops people from abusing it and adding gibberish, just to pass the form validation), and I also see why this issue should be closed 😓  ...I still think it's technically wrong, but it does have merit to leave things as is at this stage of our project.

**Comment by BWPanda** [comment]
Maybe we should let users choose what happens on their sites. E.g. have a setting that specifies whether images with no alt text get an empty attribute added, or none at all...

**Comment by klonos** [comment]
> Maybe we should let users choose what happens on their sites. E.g. have a setting that specifies whether images with no alt text get an empty attribute added, or none at all...

Sounds ideal, but the usual problem we are having in such cases is where should these settings live.

**Comment by olafgrabienski** [comment]
> Maybe we should let users choose what happens on their sites. E.g. have a setting that specifies whether images with no alt text get an empty attribute added, or none at all...

The issue isn't about "arbitrarily" adding empty `alt` attributes but only when a content editor leaves the respective field empty. What's wrong with providing the empty attribute in this case? It seems to be the most reasonable behavior then. Don't see the need for a setting here.

**Comment by klonos** [comment]
> What's wrong with providing the empty attribute in this case?

As I've mentioned previously, adding `alt=""` instructs assistive technology to "skip" images and treat them as "decorative" (a border or background), when in fact they may be the main image of the article. This is an actual problem for people that are using assistive technology, and although automatic tools that validate accessibility may be tricked, a manual accessibility assessment will fail.

**Comment by olafgrabienski** [comment]
> adding `alt=""` instructs assistive technology to "skip" images and treat them as "decorative"

I see, that may be a problem, but based on a decision of an editor. In that case, would the issue be fixed or mitigated by not adding the `alt`attribute at all? I don't think so, at least not without introducing other issues.

**Comment by klonos** [comment]
> ...at least not without introducing other issues.

Yup. This is a tough nut to crack. Ensuring proper alt text relies on the assumption that humans (the content editors) will be doing the right thing (adding proper, meaningful alt text) ...even in Drupal 8/9 where the alt text has been made a requirement, content editors seem to be adding gibberish, nonsensical text to it, just to "bypass" the form validation. So that is not a solution either 🤷 ...the only way that this could work is with AI-based validation, which is not available to us 😅 

At this point, as discussed earlier, we have the following options:

1. Leave things as they are; although we know that adding `alt=""` at all times when the text field is left empty is not the proper thing to do - we simply rely on content editors to do the right thing. Regardless of that, pages with uploaded images created in Backdrop will pass accessibility tests (even if not actually properly accessible).
2. Make it so that `alt=""` is not added automatically, if the respective field is left empty. This would "technically" be correct, but it will cause Backdrop to get a bad accessibility rating.

At this point, I should also mention that WCAG discussed the cases where either the file name or portion of the page text is automatically added to the alt attribute; these are also considered accessibility fails, and not proper solutions.

**Comment by indigoxela** [comment]
Good web accessibility can be hard to achieve if it comes to images. Has always been, will always be.

Personally, I don't think there's a proper _technical_ solution to solve this. I do belief, that our current solution is the best variant.

If editors don't provide an alt text, an empty attribute gets added, which signals "this image is just decoration". That seems better than no alt attribute (giving no hint at all to the readers). Or even worse - misleading and unhelpful alternative descriptions, which frustrate both, the editors (what should I write?) and the readers (what does that mean?).

**Comment by BWPanda** [comment]
Here's a weird (made-up) analogy:

I ask my wife to leave some chocolate cookies for me in a bowl on the table. If I arrive and there's a bowl on the table with cookies in it, I know my wife did what I asked and loves me; yay! If I arrive and there's no bowl on the table, I figure my wife just forgot to do as I asked; no biggie. However if I arrive and there's an empty bowl on the table, then my wife knew what I asked, but decided not to leave me any cookies; ☹️

Take from that what you will, but I think no alt attribute ('forgetfulness') is better than an empty alt attribute ('willful disobedience').

**Comment by docwilmot** [comment]
> I ask my wife to leave some chocolate cookies for me in a bowl on the table. If I arrive and there's a bowl on the table with cookies in it, I know my wife did what I asked and loves me; yay! If I arrive and there's no bowl on the table, I figure my wife just forgot to do as I asked; no biggie. However if I arrive and there's an empty bowl on the table, then my wife knew what I asked, but decided not to leave me any cookies; ☹️

😆 

Its more like youre running a restaurant; and the National Restaurant Association requires that everyone be provided a bowl (for cookies maybe) on every table. Even if you personally think empty bowls are rude, your customers hardly notice, they dont mind, and the Association is happy.

I vote for empty bowls.

**Comment by oadaeh** [comment]
> However if I arrive and there's an empty bowl on the table, then my wife knew what I asked, but decided not to leave me any cookies
What if she did what you asked, but someone else found the bowl and ate all the cookies before you arrive?

**Comment by oadaeh** [comment]
> At this point, as discussed earlier, we have the following options:

What about a third option: Improve the description, so people understand the importance of filling in the field and implications of not filling it in. It could be that most people view it as unnecessary, simply because they don't understand.

**Comment by klonos** [comment]
>What about a third option: Improve the description, so people understand the importance of filling in the field and implications of not filling it in. It could be that most people view it as unnecessary, simply because they don't understand.

I like that idea, but we'll need to come up with something short and to-the-point.

**Comment by laryn** [comment]
I agree with @indigoxela's [comment](https://github.com/backdrop/backdrop-issues/issues/4551#issuecomment-676297920). 

I'll also add that I think some screen readers will read the filename if there is no alt text (and no blank alt value), which is generally not good. I think if the editor decides the image isn't worth adding alt text to, adding a blank value is the best technical option in terms of accessibility. Adding description text to say that is going to happen is not necessarily a bad idea as long as it is "short and to-the-point."

**Comment by klonos** [comment]
> ...if the editor decides the image isn't worth adding alt text to

I don't think that that's an "active" decision. I refuse to believe that people actually know what the purpose of this text is, yet actively choose to ignore it. We need to find some nice words to basically say: "Blind people that cannot see images, but they can read text. Please add some text for them here."

**Comment by BWPanda** [comment]
All these differing opinions lead me to suggest, again, that a setting for this behaviour would fix all use cases: those who want no alt text to result in no alt attribute can have that, and those who want no alt text to result in an empty alt attribute can have that.

We can also add help text too, as mentioned above.

**Comment by olafgrabienski** [comment]
> What about a third option: Improve the description, so people understand the importance of filling in the field and implications of not filling it in.

For reference, this is the current description for "Alternate text".

`This text will be used by screen readers, search engines, or when the image cannot be loaded.`

That's not *so* bad, in my opinion, but let's see if we can improve it, e.g. by not only describing but inviting. First attempts:

`Provide some text for people who use screen readers, search engines, and situations when the image cannot be loaded.`

`Describe what would be useful for people who cannot see the image. The text will be used by screen readers, search engines, or when the image cannot be loaded. `

**Comment by indigoxela** [comment]
> ... current description ... Not so bad, in my opinion

I actually belief that this description pretty good.

> I refuse to believe that people actually know what the purpose of this text is, yet actively choose to ignore it.

It should be noted that it can be really hard to figure out a good alternate text. I don't think that editors don't fill out the alt attribute because they don't care. I *do* care and more than once I had to give up after quite some time head scratching - and left the field empty (= image is decorative). So let's be fair to editors.
